File: scripts/generate_audio.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#   "requests",
# ]
# ///
"""
Generate Japanese pronunciation MP3s for all terms and kimarite cards
using Google Cloud Text-to-Speech Neural2 voices.

Usage:
    GOOGLE_TTS_API_KEY=your_key uv run scripts/generate_audio.py

Output: public/audio/{deck}/{slug}.mp3
Skips files that already exist — safe to re-run when new cards are added.
"""

import base64
import json
import os
import re
import sys
import time
from pathlib import Path

import requests

API_KEY = os.environ.get("GOOGLE_TTS_API_KEY")
if not API_KEY:
    print("Error: GOOGLE_TTS_API_KEY environment variable is not set", file=sys.stderr)
    sys.exit(1)

VOICE_NAME = "ja-JP-Neural2-B"  # Natural female voice
API_URL = f"https://texttospeech.googleapis.com/v1/text:synthesize?key={API_KEY}"

ROOT = Path(__file__).parent.parent

DECKS = [
    ("terms",    ROOT / "data/seed/terms.json"),
    ("kimarite", ROOT / "data/seed/kimarite.json"),
]


def term_slug(term: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", term.lower()).strip("-")
    return slug


def synthesize(japanese: str) -> bytes:
    response = requests.post(
        API_URL,
        json={
            "input": {"text": japanese},
            "voice": {"languageCode": "ja-JP", "name": VOICE_NAME},
            "audioConfig": {"audioEncoding": "MP3", "speakingRate": 0.9},
        },
        timeout=15,
    )
    response.raise_for_status()
    return base64.b64decode(response.json()["audioContent"])


generated = 0
skipped = 0

for deck_slug, deck_path in DECKS:
    cards = json.loads(deck_path.read_text())
    out_dir = ROOT / "public" / "audio" / deck_slug
    out_dir.mkdir(parents=True, exist_ok=True)

    for card in cards:
        filename = f"{term_slug(card['term'])}.mp3"
        filepath = out_dir / filename

        if filepath.exists():
            print(f"  skip  {deck_slug}/{filename}")
            skipped += 1
            continue

        try:
            audio = synthesize(card["japanese"])
            filepath.write_bytes(audio)
            print(f"  gen   {deck_slug}/{filename}  ({len(audio):,} bytes)")
            generated += 1
        except requests.HTTPError as e:
            print(f"  error {deck_slug}/{filename}: {e.response.status_code} {e.response.text}", file=sys.stderr)
        except Exception as e:
            print(f"  error {deck_slug}/{filename}: {e}", file=sys.stderr)

        time.sleep(0.2)  # stay well within rate limits

print(f"\nDone. {generated} generated, {skipped} skipped.")

File: scripts/generate_audio.ts
/**
 * Generates Japanese pronunciation MP3s for all terms and kimarite cards
 * using Google Cloud Text-to-Speech Neural2 voices.
 *
 * Usage:
 *   GOOGLE_TTS_API_KEY=your_key npm run generate:audio
 *
 * Output: public/audio/{deck}/{slug}.mp3
 * Skips files that already exist — safe to re-run when new cards are added.
 */

import fs from "node:fs";
import path from "node:path";
import terms from "../data/seed/terms.json" assert { type: "json" };
import kimarite from "../data/seed/kimarite.json" assert { type: "json" };

const API_KEY = process.env.GOOGLE_TTS_API_KEY;
if (!API_KEY) {
  console.error("Missing GOOGLE_TTS_API_KEY environment variable");
  process.exit(1);
}

const VOICE_NAME = "ja-JP-Neural2-B"; // Natural female voice
const API_URL = `https://texttospeech.googleapis.com/v1/text:synthesize?key=${API_KEY}`;

function termSlug(term: string): string {
  return term.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
}

async function synthesize(text: string): Promise<Buffer> {
  const res = await fetch(API_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      input: { text },
      voice: { languageCode: "ja-JP", name: VOICE_NAME },
      audioConfig: { audioEncoding: "MP3", speakingRate: 0.9 }
    })
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`TTS API error ${res.status}: ${err}`);
  }

  const data = await res.json() as { audioContent: string };
  return Buffer.from(data.audioContent, "base64");
}

const decks: Array<{ slug: string; cards: Array<{ term: string; japanese: string }> }> = [
  { slug: "terms", cards: terms },
  { slug: "kimarite", cards: kimarite }
];

let generated = 0;
let skipped = 0;

for (const deck of decks) {
  const dir = path.join("public", "audio", deck.slug);
  fs.mkdirSync(dir, { recursive: true });

  for (const card of deck.cards) {
    const filename = `${termSlug(card.term)}.mp3`;
    const filepath = path.join(dir, filename);

    if (fs.existsSync(filepath)) {
      console.log(`  skip  ${deck.slug}/${filename}`);
      skipped++;
      continue;
    }

    try {
      const audio = await synthesize(card.japanese);
      fs.writeFileSync(filepath, audio);
      console.log(`  gen   ${deck.slug}/${filename}  (${audio.length} bytes)`);
      generated++;
    } catch (err) {
      console.error(`  error ${deck.slug}/${filename}: ${err}`);
    }

    // Avoid hitting rate limits
    await new Promise((r) => setTimeout(r, 200));
  }
}

console.log(`\nDone. ${generated} generated, ${skipped} skipped.`);
